File: packages/shared-types/src/demographic.ts
// ============================================================
// CORE DEMOGRAPHIC TYPES
// ============================================================

/** Gender enumeration */
export type Gender = 'male' | 'female';

/** Age group representation (single year or range) */
export interface AgeGroup {
  readonly min: number;
  readonly max: number; // For single-year: min === max
}

/** Geographic region identifier */
export interface Region {
  readonly id: string;
  readonly name: string;
  readonly parentId?: string; // For hierarchical regions (country → state → city)
}

/** Population cohort: count of people by age, gender, region */
export interface Cohort {
  readonly age: number;        // Single year of age (0-120)
  readonly gender: Gender;
  readonly regionId: string;
  readonly count: number;
}

/** Population metadata */
export interface PopulationMetadata {
  readonly totalPopulation: number;
  readonly medianAge: number;
  readonly maleCount: number;
  readonly femaleCount: number;
}

/** Population: Collection of cohorts for a point in time */
export interface Population {
  readonly scenarioId: string;
  readonly year: number;
  readonly cohorts: readonly Cohort[];
  readonly metadata: PopulationMetadata;
}

// ============================================================
// DEMOGRAPHIC TABLES
// ============================================================

/** Mortality rate entry by age and gender */
export interface MortalityRate {
  readonly age: number;
  readonly male: number;    // Probability of death (0-1)
  readonly female: number;  // Probability of death (0-1)
}

/** Mortality rates table for a region and year */
export interface MortalityTable {
  readonly regionId: string;
  readonly year: number;
  readonly rates: readonly MortalityRate[];
}

/** Fertility rate entry by mother's age */
export interface FertilityRate {
  readonly age: number;     // Mother's age (typically 15-49)
  readonly rate: number;    // Births per woman per year
}

/** Fertility rates table for a region and year */
export interface FertilityTable {
  readonly regionId: string;
  readonly year: number;
  readonly rates: readonly FertilityRate[];
  readonly sexRatioAtBirth: number; // Males per 100 females (typically ~105)
}

// ============================================================
// MIGRATION
// ============================================================

/** Standard migration age profile types */
export type MigrationProfileType = 
  | 'labor'      // Young adults (20-35 peak)
  | 'family'     // Parents with children
  | 'refugee'    // Crisis migration (all ages)
  | 'retirement' // Elderly (60+ peak)
  | 'student';   // Young adults (18-25 peak)

/** Migration flow between regions or in/out of system */
export interface MigrationFlow {
  readonly year: number;
  readonly fromRegionId: string | null;  // null = immigration from outside
  readonly toRegionId: string | null;    // null = emigration to outside
  readonly count: number;
  readonly profile: MigrationProfileType | readonly number[]; // Profile type or custom age distribution
  readonly genderRatio: number; // Males per 100 migrants (50 = equal)
}

// ============================================================
// SHOCKS (SCENARIO MODIFIERS)
// ============================================================

/** Shock type */
export type ShockType = 'mortality' | 'fertility' | 'migration';

/** Shock modifier - applies to demographic rates */
export interface Shock {
  readonly id: string;
  readonly name: string;
  readonly description?: string;
  readonly type: ShockType;
  readonly startYear: number;
  readonly endYear: number;
  readonly targetRegions: readonly string[] | 'all';
  readonly targetGenders: readonly Gender[] | 'all';
  readonly targetAges: AgeGroup | 'all';
  readonly modifier: number; // Multiplier (1.5 = 50% increase, 0.8 = 20% decrease)
}

// ============================================================
// SCENARIOS
// ============================================================

/** Complete scenario definition */
export interface Scenario {
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly baseYear: number;
  readonly endYear: number;
  readonly regions: readonly string[];
  readonly shocks: readonly Shock[];
  readonly createdAt: string;  // ISO 8601
  readonly updatedAt: string;  // ISO 8601
}

/** Scenario summary (for listings) */
export interface ScenarioSummary {
  readonly id: string;
  readonly name: string;
  readonly baseYear: number;
  readonly endYear: number;
  readonly regionCount: number;
  readonly shockCount: number;
  readonly createdAt: string;
}

// ============================================================
// PROJECTION RESULTS
// ============================================================

/** Single year projection result */
export interface ProjectionYear {
  readonly year: number;
  readonly totalPopulation: number;
  readonly births: number;
  readonly deaths: number;
  readonly netMigration: number;
  readonly naturalChange: number; // births - deaths
  readonly growthRate: number;    // Percentage change from previous year
}

/** Complete projection result */
export interface ProjectionResult {
  readonly scenarioId: string;
  readonly computedAt: string;    // ISO 8601
  readonly computeTimeMs: number;
  readonly baseYear: number;
  readonly endYear: number;
  readonly years: readonly ProjectionYear[];
}

/** Projection progress update */
export interface ProjectionProgress {
  readonly scenarioId: string;
  readonly currentYear: number;
  readonly totalYears: number;
  readonly percentComplete: number;
  readonly estimatedRemainingMs?: number;
}

// ============================================================
// AGGREGATION
// ============================================================

/** Options for aggregating population data */
export interface AggregationOptions {
  readonly ageGroupSize?: number;  // Group into N-year buckets (5, 10, etc.)
  readonly byGender?: boolean;
  readonly byRegion?: boolean;
}

/** Aggregated population data point */
export interface AggregatedPopulation {
  readonly ageGroupStart: number;
  readonly ageGroupEnd: number;
  readonly gender?: Gender;
  readonly regionId?: string;
  readonly count: number;
  readonly percentage: number;
}

// ============================================================
// UTILITY TYPES
// ============================================================

/** Create a mutable version of a readonly type */
export type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

/** Create a new scenario (without id, timestamps) */
export type NewScenario = Omit<Scenario, 'id' | 'createdAt' | 'updatedAt'>;

/** Create a new shock (without id) */
export type NewShock = Omit<Shock, 'id'>;

// ============================================================
// HELPER FUNCTIONS
// ============================================================

/** Create an age group for a single year */
export function singleYearAge(age: number): AgeGroup {
  return { min: age, max: age };
}

/** Create an age group for a range */
export function ageRange(min: number, max: number): AgeGroup {
  if (min > max) {
    throw new Error('Invalid age range: min cannot be greater than max');
  }
  return { min, max };
}

/** Check if an age falls within an age group */
export function isAgeInGroup(age: number, group: AgeGroup | 'all'): boolean {
  if (group === 'all') {
    return true;
  }
  return age >= group.min && age <= group.max;
}

/** Calculate total population from cohorts */
export function totalPopulation(cohorts: readonly Cohort[]): number {
  return cohorts.reduce((sum, cohort) => sum + cohort.count, 0);
}

/** Calculate median age from cohorts */
export function medianAge(cohorts: readonly Cohort[]): number {
  if (cohorts.length === 0) {
    return 0;
  }

  const total = totalPopulation(cohorts);
  if (total === 0) {
    return 0;
  }

  // Sort cohorts by age
  const sorted = [...cohorts].sort((a, b) => a.age - b.age);
  
  // Find median age (the age where cumulative count reaches half)
  const half = total / 2;
  let cumulative = 0;
  
  for (const cohort of sorted) {
    cumulative += cohort.count;
    if (cumulative >= half) {
      return cohort.age;
    }
  }
  
  const lastCohort = sorted[sorted.length - 1];
  return lastCohort?.age ?? 0;
}

File: worker/src/types/demographic.rs
//! Core demographic types

use serde::{Deserialize, Serialize};

/// Gender enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Gender {
    Male,
    Female,
}

/// Age group representation
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgeGroup {
    pub min: u32,
    pub max: u32,
}

impl AgeGroup {
    pub fn single_year(age: u32) -> Self {
        Self { min: age, max: age }
    }

    pub fn range(min: u32, max: u32) -> Self {
        Self { min, max }
    }

    pub fn contains(&self, age: u32) -> bool {
        age >= self.min && age <= self.max
    }
}

/// Geographic region identifier
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Region {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
}

/// Cohort: Population count by age, gender, and region
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Cohort {
    pub age: u32,
    pub gender: Gender,
    pub region_id: String,
    pub count: f64,
}

/// Population metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PopulationMetadata {
    pub total_population: f64,
    pub male_population: f64,
    pub female_population: f64,
    pub median_age: f64,
    pub dependency_ratio: f64,
}

/// Population: Complete demographic state at a point in time
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Population {
    pub cohorts: Vec<Cohort>,
    pub metadata: PopulationMetadata,
}

impl Population {
    /// Create a new population from cohorts, computing metadata automatically
    pub fn from_cohorts(cohorts: Vec<Cohort>) -> Self {
        let metadata = Self::compute_metadata(&cohorts);
        Self { cohorts, metadata }
    }

    fn compute_metadata(cohorts: &[Cohort]) -> PopulationMetadata {
        let mut total = 0.0;
        let mut male = 0.0;
        let mut female = 0.0;
        let mut young = 0.0;  // 0-14
        let mut old = 0.0;    // 65+
        
        for c in cohorts {
            total += c.count;
            match c.gender {
                Gender::Male => male += c.count,
                Gender::Female => female += c.count,
            }
            if c.age < 15 {
                young += c.count;
            } else if c.age >= 65 {
                old += c.count;
            }
        }

        let working_age = total - young - old;
        let dependency_ratio = if working_age > 0.0 {
            (young + old) / working_age * 100.0
        } else {
            0.0
        };

        PopulationMetadata {
            total_population: total,
            male_population: male,
            female_population: female,
            median_age: Self::compute_median_age(cohorts, total),
            dependency_ratio,
        }
    }

    fn compute_median_age(cohorts: &[Cohort], total: f64) -> f64 {
        if total == 0.0 {
            return 0.0;
        }

        let mut sorted: Vec<_> = cohorts.iter().collect();
        sorted.sort_by_key(|c| c.age);

        let half = total / 2.0;
        let mut cumulative = 0.0;

        for cohort in sorted {
            cumulative += cohort.count;
            if cumulative >= half {
                return cohort.age as f64;
            }
        }

        0.0
    }
}

/// Mortality rate for a specific age
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MortalityRate {
    pub age: u32,
    pub male: f64,
    pub female: f64,
}

/// Mortality table: Death probabilities by age and gender
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MortalityTable {
    pub region_id: String,
    pub year: u32,
    pub rates: Vec<MortalityRate>,
}

impl MortalityTable {
    /// Get mortality rate for a specific age and gender
    pub fn get_rate(&self, age: u32, gender: Gender) -> f64 {
        self.rates
            .iter()
            .find(|r| r.age == age)
            .map(|r| match gender {
                Gender::Male => r.male,
                Gender::Female => r.female,
            })
            .unwrap_or(1.0) // Default to 100% mortality for ages not in table
    }
}

/// Fertility rate for a specific age
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FertilityRate {
    pub age: u32,
    pub rate: f64,
}

/// Fertility table: Birth rates by mother's age
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FertilityTable {
    pub region_id: String,
    pub year: u32,
    pub rates: Vec<FertilityRate>,
    /// Males per 100 females at birth (typically ~105)
    pub sex_ratio_at_birth: f64,
}

impl FertilityTable {
    /// Get fertility rate for a specific age
    pub fn get_rate(&self, age: u32) -> f64 {
        self.rates
            .iter()
            .find(|r| r.age == age)
            .map(|r| r.rate)
            .unwrap_or(0.0)
    }
}

/// Projection result for a single year
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectionYear {
    pub year: u32,
    pub population: Population,
    pub births: f64,
    pub deaths: f64,
    pub net_migration: f64,
}

/// Complete projection result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectionResult {
    pub scenario_id: String,
    pub computed_at: String,
    pub compute_time_ms: u64,
    pub years: Vec<ProjectionYear>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_age_group_contains() {
        let group = AgeGroup::range(18, 65);
        assert!(group.contains(18));
        assert!(group.contains(40));
        assert!(group.contains(65));
        assert!(!group.contains(17));
        assert!(!group.contains(66));
    }

    #[test]
    fn test_population_metadata() {
        let cohorts = vec![
            Cohort { age: 10, gender: Gender::Male, region_id: "CZ".into(), count: 100.0 },
            Cohort { age: 30, gender: Gender::Female, region_id: "CZ".into(), count: 100.0 },
            Cohort { age: 70, gender: Gender::Male, region_id: "CZ".into(), count: 50.0 },
        ];
        
        let pop = Population::from_cohorts(cohorts);
        
        assert_eq!(pop.metadata.total_population, 250.0);
        assert_eq!(pop.metadata.male_population, 150.0);
        assert_eq!(pop.metadata.female_population, 100.0);
    }
}

